interfaces/apparmor/spec.go

```diff
@@ -567,6 +567,61 @@ func snippetFromLayout(layout *snap.Layout) string {
 	return fmt.Sprintf("# Layout path: %s\n# (no extra permissions required for symlink)", mountPoint)
 }
 
+// emitUserEnsureDir creates an apparmor snippet that permits snap-update-ns to create
+// missing directories for the calling user according to the provided ensure directory spec.
+func emitUserEnsureDir(spec *Specification, ifaceName string, ensureDirSpec *interfaces.EnsureDirSpec) {
+	ensureDir := filepath.Clean(ensureDirSpec.EnsureDir)
+	mustExistDir := filepath.Clean(ensureDirSpec.MustExistDir)
+	if ensureDir == mustExistDir {
+		return
+	}
+
+	replacePrefixHome := func(path string) string {
+		if strings.HasPrefix(path, "$HOME") {
+			return strings.Replace(path, "$HOME", "@{HOME}", -1)
+		}
+		return path
+	}
+	appArmorDir := func(path string) string {
+		if path != "/" {
+			path = path + "/"
+		}
+		return path
+	}
+	emit := spec.AddUpdateNSf
+
+	// Create entry for MustExistDir
+	iter, err := strutil.NewPathIterator(ensureDir)
+	if err != nil {
+		panic(err)
+	}
+	for iter.Next() {
+		if iter.CurrentCleanPath() == mustExistDir {
```

Comment:
why do we need a Clean variant given that we clean the inputs above?

Reply:
Yes, `mustExistDir` will be clean, and the iterator will output `<mustExistDir>/`

Reply:
but ensureDir is also clean, no? I'm still unsure why the need for CurrentCleanPath ?

Reply:
actually NewPathIterator double checks that, so I'm still not sure why this change to PathIterator

Reply:
now I see, I think I had the wrong mental model of what PathIterator does, anway a different name would be better

Reply:
Apologies if I am missing something.

The problem in my view is this:
![Screenshot from 2023-10-26 14-47-27](https://github.com/snapcore/snapd/assets/5872705/39a1a903-eb6d-4e03-b5fe-bbc980663fea)

Only the full path is clean (does not have final slash), but for all directories except the full path, it returns path with "/".